server/src/main/java/com/hedera/block/server/persistence/cache/LRUCache.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.block.server.persistence.cache;
+
+import com.hedera.block.protos.BlockStreamServiceGrpcProto;
+import jakarta.inject.Singleton;
+
+import java.util.Collections;
+import java.util.Map;
+
+/**
+ * An LRU cache implementation which uses a custom LinkedHashMap to store blocks in memory and
+ * evict the least recently used block when the cache size exceeds a specified limit.
+ */
+@Singleton
+public class LRUCache implements BlockCache<BlockStreamServiceGrpcProto.Block> {
+
+    private final Map<Long, BlockStreamServiceGrpcProto.Block> cache;
+
+    /**
+     * Constructor for the LRUCache class.
+     *
+     * @param maxEntries the maximum number of entries in the cache
+     */
+    public LRUCache(final long maxEntries) {
+        final System.Logger LOGGER = System.getLogger(getClass().getName());
+        LOGGER.log(System.Logger.Level.INFO, "Creating LRUCache with maxEntries: " + maxEntries);
+
+        cache = Collections.synchronizedMap(new BNLinkedHashMap<>(maxEntries));
```

Comment:
This adds a lock to every access.  We would be _much_ better off using a concurrent lock-free implementation.
Perhaps something based on a ConcurrentSkipListMap, for instance, with the "LRU" removal element _separate_ from the storage element and a _composition_ of the two forming the cache object.

Reply:
I removed the cache